cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,316 @@
+	"sync"
+	"time"
+
+	"github.com/digitalocean/godo"
+	"github.com/google/go-cmp/cmp"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+const (
+	// Interval of synchronizing service status from apiserver.
+	serviceSyncPeriod = 30 * time.Second
+	// Frequency at which the firewall controller runs.
+	firewallReconcileFrequency = 5 * time.Minute
+)
+
+var (
+	// Sentinel errors make it easier to test.
+	errFailedToAddInboundRules        = errors.New("failed to add new firewall inbound rules")
+	errFailedToCreateFirewallOnUpdate = errors.New("failed to create firewall after found missing on update")
+	errFailedToListServices           = errors.New("failed to list services")
+	errFailedToRetrieveFirewallByID   = errors.New("failed to retrieve firewall by ID")
+	errFailedToRetrieveFirewallList   = errors.New("failed to retrieve list of firewalls from DO API")
+)
+
+// firewallCache stores a cached firewall and mutex to handle concurrent access.
+type firewallCache struct {
+	mu       *sync.RWMutex // protects firewall.
+	firewall *godo.Firewall
+}
+
+// firewallManagerOp manages the interaction with the DO Firewalls API.
+type firewallManagerOp struct {
+	client             *godo.Client
+	fwCache            firewallCache
+	workerFirewallName string
+	workerFirewallTags []string
+}
+
+// FirewallManager retrieves and stores firewall representations.
+type FirewallManager interface {
+	// Get returns the current public access firewall representation (i.e., the DO Firewall object)
+	// if it exists and nil if it does not exist.
+	Get(ctx context.Context) (*godo.Firewall, error)
+
+	// Set applies the given inbound rules to the public access firewall.
+	Set(ctx context.Context, inboundRules []godo.InboundRule) error
+}
+
+// FirewallController helps to keep cloud provider service firewalls in sync.
+type FirewallController struct {
+	kubeClient         clientset.Interface
+	client             *godo.Client
+	workerFirewallTags []string
+	serviceLister      corelisters.ServiceLister
+	fwManager          FirewallManager
+}
+
+// NewFirewallController returns a new firewall controller to reconcile public access firewall state.
+func NewFirewallController(
+	ctx context.Context,
+	kubeClient clientset.Interface,
+	client *godo.Client,
+	serviceInformer coreinformers.ServiceInformer,
+	fwManager FirewallManager,
+	workerFirewallTags []string,
+) *FirewallController {
+	fc := &FirewallController{
+		kubeClient:         kubeClient,
+		client:             client,
+		workerFirewallTags: workerFirewallTags,
+		fwManager:          fwManager,
+	}
+
+	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				fc.ensureReconciledFirewall(ctx)
```

Comment:
The `ctx` is set to `context.Background()` where `NewFirewallController()` is called, meaning it'll never be used for cancelation. I think we should set a context with a reasonable timeout for each of the `*Func` handlers, perhaps something on the order of 90 seconds?

(We can then also remove the `context.Context` type from the `NewFirewallController` signature.)